hail/src/main/resources/include/hail/Region2.h

```diff
@@ -0,0 +1,99 @@
+#ifndef HAIL_REGION2_H
+#define HAIL_REGION2_H 1
+
+#include <memory>
+#include <vector>
+#include <utility>
+#include "hail/Upcalls.h"
+#include "hail/NativeStatus.h"
+
+namespace hail {
+
+class RegionPool {
+  static constexpr ssize_t block_size = 64*1024;
+  static constexpr ssize_t block_threshold = 4096;
+
+  public:
+    class Region {
+      public:
+        class SharedPtr {
+          private:
+            Region * region_;
+            void clear();
+          public:
+            SharedPtr(Region * region) : region_(region) {
+              if (region_ != nullptr) { ++(region_->references_); }
+            }
+            SharedPtr(const SharedPtr &ptr) : SharedPtr(ptr.region_) { }
+            SharedPtr(SharedPtr &&ptr) : region_(nullptr) {
+              region_ = ptr.region_;
+              ptr.region_ = nullptr;
+            }
+            ~SharedPtr() { clear(); }
+
+            void swap(SharedPtr &other) { std::swap(region_, other.region_); }
+            SharedPtr& operator=(SharedPtr other) {
+              swap(other);
+              return *this;
+            }
+            SharedPtr& operator=(std::nullptr_t) noexcept {
+              clear();
+              return *this;
+            }
+            inline Region * get() { return region_; }
```

Comment:
`inline` is redundant on an in-class method definition. C++ is already keyword soup, so I prefer to leave it off, and I would definitely like to be consistent (otherwise readers may wonder what the significance is).